source/Test.cpp

```diff
@@ -192,12 +274,81 @@ void Test::Step(Context &context, UI &menuPanels, UI &gamePanels, PlayerInfo &pl
 		return;
 	}
 	
-	const TestStep &stepToRun = steps[context.stepToRun];
+	// All processing was done just before this step started.
+	context.unprocessedInput = false;
+	context.branchesSinceGameStep.clear();
 	
-	// Exit with error on a failing testStep.
-	const string &stepTypeName = STEPTYPE_TO_TEXT.at(stepToRun.stepType);
-	
-	Fail("Test step " + to_string(context.stepToRun) + " (" + stepTypeName + ") failed");
+	bool continueGameLoop = false;
+	while(context.stepToRun < steps.size() && !continueGameLoop)
+	{
+		// Fail if we encounter a watchdog timeout
+		if(context.watchdog == 1)
+			Fail(context, player, "watchdog timeout");
+		else if(context.watchdog > 1)
+			--(context.watchdog);
+		
+		const TestStep &stepToRun = steps[context.stepToRun];
+		switch(stepToRun.stepType)
+		{
+			case TestStep::Type::APPLY:
+				stepToRun.conditions.Apply(player.Conditions());
+				++(context.stepToRun);
+				break;
+			case TestStep::Type::ASSERT:
+				if(!stepToRun.conditions.Test(player.Conditions()))
+					Fail(context, player, "asserted false");
```

Comment:
Should we pass the conditionset that failed rather than a failure message in this case? This would allow us to print it and only the associated player conditions.

(We can defer this to a follow-up if desired)

Reply:
Let's defer this to follow-up. I'm okay with the idea of only printing the conditions used in the (whole) test, but only printing the conditions from a single step could provide an incomplete picture (if variables around the step are relevant).